src/security/builtin_plugins/authentication/src/auth_utils.c

```diff
@@ -0,0 +1,1530 @@
+
+    name = X509_get_subject_name(cert);
+    if (!name) {
+        char *msg = get_openssl_error_message();
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "X509_get_subject_name failed : %s", msg);
+        ddsrt_free(msg);
+        goto err_get_subject;
+    }
+
+    /* TODO: check if this is the correct format of the subject name: check spec */
+    X509_NAME_print_ex(bio, name, 0, XN_FLAG_RFC2253);
+
+    sz = (size_t)BIO_get_mem_data(bio, &pmem);
+    subject = ddsrt_malloc( sz + 1);
+
+    if (BIO_gets(bio, subject, (int32_t)sz + 1) < 0) {
+        char *msg = get_openssl_error_message();
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "X509_get_subject_name failed : %s", msg);
+        ddsrt_free(msg);
+        ddsrt_free(subject);
+        subject = NULL;
+    }
+
+    BIO_free(bio);
+
+    return subject;
+
+err_get_subject:
+    BIO_free(bio);
+err_bio_alloc:
+    return NULL;
+}
+
+dds_time_t
+get_certificate_expiry(
+    X509 *cert)
+{
+    dds_time_t expiry = DDS_TIME_INVALID;
+    ASN1_TIME *ans1;
+
+    assert(cert);
+
+    ans1 = X509_get_notAfter(cert);
+    if (ans1 != NULL) {
+        int days;
+        int seconds;
+        if (ASN1_TIME_diff(&days, &seconds, NULL, ans1) != 0) {
+            static const dds_duration_t secs_per_day = 86400;
+            dds_duration_t delta = ((dds_duration_t)seconds + ((dds_duration_t)days * secs_per_day)) * DDS_NSECS_IN_SEC;
+            expiry = dds_time() + delta;
+        }
+    }
+
+    return expiry;
+}
+
+int32_t
+get_subject_name_DER_encoded(
+        const X509 *cert,
+        unsigned char **buffer,
+        DDS_Security_SecurityException *ex)
+{
+    X509_NAME *name;
+    unsigned char *tmp = NULL;
+    int32_t sz = -1;
+
+    assert(cert);
+    assert(buffer);
+
+    name = X509_get_subject_name((X509 *)cert);
+    if (name) {
+        sz = i2d_X509_NAME(name, &tmp);
+        if (sz > 0) {
+            *buffer = ddsrt_malloc((size_t)sz);
+            memcpy(*buffer, tmp, (size_t)sz);
+            OPENSSL_free(tmp);
+        } else if (sz < 0) {
+            char *msg = get_openssl_error_message();
+            DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "i2d_X509_NAME failed : %s", msg);
+            ddsrt_free(msg);
+        }
+    } else {
+        char *msg = get_openssl_error_message();
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "X509_get_subject_name failed : %s", msg);
+        ddsrt_free(msg);
+    }
+
+    return sz;
+}
+
+
+static DDS_Security_ValidationResult_t
+check_key_type_and_size(
+    EVP_PKEY *key,
+    int isPrivate,
+    DDS_Security_SecurityException *ex)
+{
+    DDS_Security_ValidationResult_t result = DDS_SECURITY_VALIDATION_OK;
```

Comment:
To me, it makes more sense to let default result of a validation routine be "not ok".

Reply:
It depends on the situation. Here, I have to write additional `else` blocks to start with DDS_SECURITY_VALIDATION_FAILED.